### ISSUE TITLE
Thread ID changes between test and *Invocation events when the test timeout set

### ISSUE THREAD

**Issue by jcdkount** [testng-team/testng#914]
Using the attached code, run the Bug.xml file.

EXPECTED: beforeInvocation, afterInvocation should always be on the same thread as the test.

ACTUAL: When you change the timeout in the annotation transformer (IAnnotationTransformer), the before and after Invocation events are on a different thread.

NOTE: In testing, I noticed that if I set the timeout by the @Test annotation, it also has the same problem, but I suspect it happening at run time might make this harder to patch, so I left in the more complex case.

Example output:

beforeInvocation ; Expected SAME Thread: 11 - pool-1-thread-1
myExampleTest ; Expected SAME Thread: 12 - TestNGInvoker-myExampleTest()
afterInvocation ; Expected SAME Thread: 11 - pool-1-thread-1

Possible ways to fix this:
- Make the thread the same even if the timeout changes.
- ITestResult or IInvokedMethod include the thread id or Thread object which is stored when the the test method is executed.

For further detail on the discussion of this issue, see https://groups.google.com/forum/#!topic/testng-users/W3lQsykKa3w

[Bug.zip](https://github.com/cbeust/testng/files/64519/Bug.zip)

**Comment by juherr** [comment]
I made some changes with TestNG 6.9.5 about timeout but I don't know if I broke things.
Could you try your sample with TestNG 6.9.4 and tell me if the issue was already present?

**Comment by juherr** [comment]
Related to #599

**Comment by cbeust** [comment]
The reason for this is that when you set a time out on a test method, that method gets run on a one-off executor so that the time out can be applied on it.

I once looked into fixing this but this changehas a lot of deep reaching consequences that I don't really have a clear resolution for.

**Comment by ricardf-cmp** [comment]
Glad to know there's an open issue for this! This is currently messing up with my logging (log4j). Is there any place where I can do some code before the test starts? Somewhat like "beforeInvocationInsideTestNGInvoker"?

**Comment by ricardf-cmp** [comment]
I've done a bit of research. I saw that on some comment you said that not using a new thread for each test run when there's a timeout would be a problem (I don't know why but hey you're the expert). What I don't get is why are you not calling the "run" method of an IHookable if there's a timeout also.

https://github.com/cbeust/testng/blob/testng-6.9.5/src/main/java/org/testng/internal/Invoker.java#L635

Does the timeout protection break SO much the whole thing?

**Comment by juherr** [comment]
> not using a new thread for each test run when there's a timeout would be a problem (I don't know why but hey you're the expert)

Because timeout is managed by a executor which is a simple way to manage timeout. But we are open to another solution.

> What I don't get is why are you not calling the "run" method of an IHookable if there's a timeout also.

It is: https://github.com/cbeust/testng/blob/testng-6.9.5/src/main/java/org/testng/internal/MethodInvocationHelper.java#L208

**Comment by ricardf-cmp** [comment]
Well, I found out a workaround. I renamed the thread name for something that contained "TestNG" and that allowed me to keep the tests under the same thread:
{code}Thread.currentThread().setName("TestNG-" + testname);{code}
@cbeust , until the fix for this bug is not comming, do you have any idea what kind of implications would this mean if I use this hack together with a timeout on @Test methods?

The problem here is that I use too many features probably heh, retryanalizer, dataproviders, timeouts, etc all at the same time heh.

**Comment by ricardf-cmp** [comment]
@juherr Sorry I was writing my last comment when you sent your answer. I double checked, that "invokeHookable(...)" is only used inside an if condition like:

{code}
        // If no timeOut, just invoke the method
        if (MethodHelper.calculateTimeOut(tm) <= 0) {
          [...]
            MethodInvocationHelper.invokeHookable(instance,
                parameterValues, hookableInstance, thisMethod, testResult);
          [...]
        }
        else {
          [...]
          MethodInvocationHelper.invokeWithTimeout(tm, instance, parameterValues, testResult);
        }
      }
{code}

And that "invokeWithTimeout" does NOT call the run method in iHookable :(

**Comment by juherr** [comment]
The call is not direct, but: "invokeWithTimeout" -> "invokeWithTimeoutWith*Executor" -> "InvokeMethodRunnable#run()#runOne()" -> "invokeHookable" -> "hookable.run(....)"

**Comment by ricardf-cmp** [comment]
I just tested out... I was using TestNG 6.9.4 but when upgraded to 6.9.5 and now it uses the IHookable run! I just compared both versions and you guys just added/fixed that. Thanks for pointing me to the right direction!

I use a log4j setup that logs into files, one per thread. Your thread magic when there's a timeout was breaking that logging set up heh.

**Comment by juherr** [comment]
> I use a log4j setup that logs into files, one per thread. Your thread magic when there's a timeout was breaking that logging set up heh.

I don't see any solution to fix this issue for the moment.

**Comment by DzmitryHumianiuk** [comment]
reference.
blocker for https://github.com/reportportal/client-java-core/issues/6

**Comment by balrajHostAnalytics** [comment]
This issue is long been pending can we expect fix in tesng 7.0.0

**Comment by juherr** [comment]
Not sure but we can try

**Comment by balrajHostAnalytics** [comment]
@juherr Can we use -Dtestng.thread.affinity=true same solution for the timeout case?

**Comment by juherr** [comment]
@balrajHostAnalytics I don't think so. If the fix is possible then it will be the default behavior.

**Comment by benzman81** [comment]
We are also face this issue mith @Before-/AfterMethod. We currently rely on the use of ThreadLocal. Using a time-out breaks this, as the @Before-/AfterMethod uses a different thread.

**Comment by asolntsev** [comment]
@cbeust @juherr @krmahadevan Hi. What stops us from fixing this? 

I can try to prepare a pull request if we all agree that _listener's "before" and "after" methods should be executed in the same thread as the test itself_. Do you agree that this is the right way to fix the problem?

**Comment by asolntsev** [comment]
The idea is very simple: run listener's "before" and "after" methods in the same thread as test itself.

**Comment by krmahadevan** [comment]
Sure. But running the `before` and `after` in the same thread as the `test` basically will pollute the `timeout` construct because the timeout is supposed to be applied only for the `test` method, but having them run as a single unit in a thread will perhaps violate this contract.

**Comment by asolntsev** [comment]
@krmahadevan There is an easy solution: _change the contract_!
Now `timeout` will be applied to the test WITH its "before" and "after" methods. It seems even logical to me.

**Comment by krmahadevan** [comment]
@asolntsev - thats a change in functionality and to me it changes the semantics of `timeout`

@juherr what do you think ?

**Comment by krmahadevan** [comment]
Also it will cause all the before and after listeners found on the classpath to be clubbed with the test and kind of deviate the functionality.

**Comment by asolntsev** [comment]
> Also it will cause all the before and after listeners found on the classpath to be clubbed with the test and kind of deviate the functionality.

And... What's the problem? I don't think it's a problem.

If you want to keep the backward compatibility, there is also an easy solution:
we could invent a new interface for listeners, like `INewTestListener` with methods `beforeTest` and `afterTest`. 
And these new methods `beforeTest` and `afterTest` will be executed in the same thread as test. 

So all people who needs the same thread will implement this new interface `INewTestListener`.

**Comment by krmahadevan** [comment]
> And... What's the problem? I don't think it's a problem.

It depends on the sort of usecases that people are handling with in their own repositories. So I believe its going to be unfair to assume that a flip will not cause any issues to anyone else.

> If you want to keep the backward compatibility, there is also an easy solution: we could invent a new interface for listeners, like INewTestListener with methods beforeTest and afterTest.
And these new methods beforeTest and afterTest will be executed in the same thread as test.

So all people who needs the same thread will implement this new interface INewTestListener.

Sure. We can consider introducing a variant of `IInvokedMethodListener` which would get invoked from within the same thread

**Comment by juherr** [comment]
I'm not sure to understand why a new interface will help.
Why not have a behavior parameter instead?

**Comment by asolntsev** [comment]
For me personally, the parameter also solves the problem.
But for a wider audience, the interface is better for 2 reasons:
1. It allows people to implement both interfaces (someone might need them both, right?)
2. for me, as an author of testing library (Selenide), it's much easier to implement the right interface than communicate to all my users that they need to change some configuration parameters.

**Comment by juherr** [comment]
AI-assisted triage: **still relevant**.

Timeout execution has evolved, but the underlying split-thread model still exists. `MethodInvocationHelper.invokeWithTimeout()` now avoids a child executor when it is already on a suitable TestNG worker thread, using a monitor thread instead. For the remaining paths it still calls `invokeWithTimeoutWithNewExecutor()`, so the test body runs on the one-off executor while the surrounding invocation-listener callbacks are invoked by the caller thread.

There is no current `GITHUB-914` regression test and I found no open/merged PR resolving the general callback/thread-affinity contract. The discussion also identified a real design choice: moving callbacks/configurations inside the timeout worker changes what the timeout measures, while preserving the current timeout scope leaves the listener/test thread split.

A regression should cover both timeout paths and first characterize the intended contract for `beforeInvocation`/test/`afterInvocation` thread identity before production code is changed. I did not execute the historical zip during this triage.
